ksql-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsLocator.java

```diff
@@ -51,20 +60,51 @@
   }
 
   @Override
-  @SuppressWarnings("deprecation")
-  public Optional<KsqlNode> locate(final Struct key) {
-    final StreamsMetadata metadata = kafkaStreams
-        .metadataForKey(stateStoreName, key, keySerializer);
+  public Optional<ActiveAndStandByNodes> locate(
+      final Struct key,
+      final Optional<Map<String, HostStatusEntity>> hostStatuses
+  ) {
+    final KeyQueryMetadata metadata = kafkaStreams
+        .queryMetadataForKey(stateStoreName, key, keySerializer);
 
-    if (metadata == StreamsMetadata.NOT_AVAILABLE) {
+    // FAil fast if Streams not ready. Let client handle it
+    if (metadata == KeyQueryMetadata.NOT_AVAILABLE) {
+      LOG.debug("Streams Metadata not available");
       return Optional.empty();
     }
 
-    final HostInfo hostInfo = metadata.hostInfo();
-    return Optional.of(asNode(hostInfo));
+    final HostInfo activeHost = metadata.getActiveHost();
+    final KsqlNode activeNode = asNode(activeHost);
+    final Set<HostInfo> standByHosts = metadata.getStandbyHosts();
+    final Map<HostInfo, KsqlNode> standByNodesMap = new HashMap<>();
+    for (HostInfo hostInfo: standByHosts) {
+      standByNodesMap.put(hostInfo, asNode(hostInfo));
+    }
+
+    LOG.debug("Active host {} , standby hosts {}", activeHost, standByHosts);
+    // mark node liveness status
+    if (hostStatuses.isPresent()) {
+      // Mark liveness of active
+      activeNode.setIsAlive(hostStatuses.get().get(activeHost.toString()).getHostAlive());
+
+      // Mark liveness of standby: Not included if dead
+      for (HostInfo standByHost: standByHosts) {
+        if (hostStatuses.get().containsKey(standByHost.toString())) {
```

Comment:
we are marking active as live/dead, but filtering out standbys.. can we just mark or filter consistently?